adobe/react-spectrum
# Fix virtualizer persisted keys with drag and drop

Depends on #6640

This fixes the virtualizer `persistedKeys` during drag and drop so that keyboard navigation works correctly and focus isn't lost. Since we only render one of the drop positions (before or after), we need to normalize the key we persist accordingly so the correct one is in the DOM and able to get focus.

I've also changed the API on Virtualizer and collection renderers in general to accept `persistedKeys` as a prop rather than a single `focusedKey`. This enables multiple keys to be persisted and is more extensible for future use cases.

Finally, based on feedback in https://github.com/adobe/react-spectrum/pull/6631#discussion_r1657916779, I've renamed the Layout `validate` method to `update` to make it easier to understand. This was originally pulled from v2 CollectionView but since we are doing a breaking change we can rename it now.

Test instructions:

* When reordering the many items story, focus shouldn't be lost when pressing and holding the down/up arrows to move quickly.
* When dropping, focus should go to the dropped item and not be lost. The focus ring should appear. It should not scroll to the top but stay on the dropped item.
* Make sure to test React 16/17/18
